core/src/test/java/org/apache/brooklyn/core/config/ConfigKeyDeprecationTest.java

```diff
@@ -0,0 +1,328 @@
+ */
+package org.apache.brooklyn.core.config;
+
+import static org.testng.Assert.assertEquals;
+
+import org.apache.brooklyn.api.entity.Entity;
+import org.apache.brooklyn.api.entity.EntityInitializer;
+import org.apache.brooklyn.api.entity.EntityLocal;
+import org.apache.brooklyn.api.entity.EntitySpec;
+import org.apache.brooklyn.api.entity.ImplementedBy;
+import org.apache.brooklyn.api.location.LocationSpec;
+import org.apache.brooklyn.api.policy.PolicySpec;
+import org.apache.brooklyn.api.sensor.EnricherSpec;
+import org.apache.brooklyn.config.ConfigKey;
+import org.apache.brooklyn.core.enricher.AbstractEnricher;
+import org.apache.brooklyn.core.entity.AbstractEntity;
+import org.apache.brooklyn.core.entity.EntityInternal;
+import org.apache.brooklyn.core.entity.internal.ConfigUtilsInternal;
+import org.apache.brooklyn.core.location.AbstractLocation;
+import org.apache.brooklyn.core.policy.AbstractPolicy;
+import org.apache.brooklyn.core.test.BrooklynAppUnitTestSupport;
+import org.apache.brooklyn.core.test.entity.TestEntity;
+import org.apache.brooklyn.entity.stock.BasicEntity;
+import org.apache.brooklyn.test.LogWatcher;
+import org.apache.brooklyn.test.LogWatcher.EventPredicates;
+import org.apache.brooklyn.util.core.config.ConfigBag;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testng.annotations.Test;
+
+import com.google.common.base.Predicate;
+import com.google.common.collect.ImmutableMap;
+
+import ch.qos.logback.classic.spi.ILoggingEvent;
+
+public class ConfigKeyDeprecationTest extends BrooklynAppUnitTestSupport {
+
+    @SuppressWarnings("unused")
+    private static final Logger log = LoggerFactory.getLogger(ConfigKeyDeprecationTest.class);
+    
+    public static final ConfigKey<String> KEY_1 = ConfigKeys.builder(String.class, "key1")
+            .deprecatedNames("oldKey1", "oldKey1b")
+            .build();
+
+    @Test
+    public void testUsingDeprecatedName() throws Exception {
+        EntityInternal entity = app.addChild(EntitySpec.create(MyBaseEntity.class)
+                .configure("oldSuperKey1", "myval"));
+        EntityInternal entity2 = app.addChild(EntitySpec.create(MyBaseEntity.class)
+                .configure("oldSuperKey1b", "myval"));
+        assertEquals(entity.config().get(MyBaseEntity.SUPER_KEY_1), "myval");
+        assertEquals(entity2.config().get(MyBaseEntity.SUPER_KEY_1), "myval");
+    }
+
+    @Test
+    public void testPrefersNonDeprecatedName() throws Exception {
+        EntityInternal entity = app.addChild(EntitySpec.create(MyBaseEntity.class)
+                .configure("superKey1", "myval")
+                .configure("oldSuperKey1", "mywrongval"));
+        assertEquals(entity.config().get(MyBaseEntity.SUPER_KEY_1), "myval");
+    }
+    
+    @Test
+    public void testPrefersFirstDeprecatedNameIfMultiple() throws Exception {
+        EntityInternal entity = app.addChild(EntitySpec.create(MyBaseEntity.class)
+                .configure("oldSuperKey1", "myval1")
```

Comment:
Maybe nicer to order these as 
```
.configure("oldSuperKey1b", "myval2")
.configure("oldSuperKey1", "myval1")
```
to demonstrate that the order in which they are configured here is not the determining factor; maybe even call `oldSuperKey1` something instead that is lexically greater than `oldSuperKey2` to demonstrate that it's not alphabetic either!